File: ios/Sources/PamNative/Modules/SQLiteModule.swift
import Foundation
import SQLite3

final class SQLiteModule: NativeModule, ClosableNativeModule {
    private let queue = DispatchQueue(label: "dev.pam.native.sqlite")
    private var databases: [String: OpaquePointer] = [:]

    func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        queue.async {
            do {
                let values = try WireMap.decode(payload)
                guard case let .text(name)? = values["database"],
                      case let .text(sql)? = values["sql"],
                      case let .text(argumentsJSON)? = values["arguments"] else {
                    throw SQLiteError("Invalid SQLite payload")
                }
                let database = try self.open(name)
                switch method {
                case "execute":
                    let arguments = try self.decodeArguments(argumentsJSON)
                    let statement = try self.prepare(database, sql)
                    defer { sqlite3_finalize(statement) }
                    try self.bind(arguments, to: statement)
                    guard sqlite3_step(statement) == SQLITE_DONE else {
                        throw SQLiteError(String(cString: sqlite3_errmsg(database)))
                    }
                    completion(.success, Data())
                case "query":
                    let arguments = try self.decodeArguments(argumentsJSON)
                    let rows = try self.query(database, sql, arguments)
                    let json = try JSONSerialization.data(withJSONObject: rows)
                    let text = String(data: json, encoding: .utf8) ?? "[]"
                    completion(.success, try WireMap.encode(["rows": .text(text)]))
                case "executeMany":
                    let argumentSets = try self.decodeArgumentSets(argumentsJSON)
                    try self.executeMany(database, sql, argumentSets)
                    completion(.success, Data())
                default:
                    throw SQLiteError("Unknown SQLite method \(method)")
                }
            } catch {
                completion(.failure, Data(error.localizedDescription.utf8))
            }
        }
    }

    private func open(_ name: String) throws -> OpaquePointer {
        if let database = databases[name] { return database }
        let root = FileManager.default.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        )[0].appendingPathComponent("pam-databases", isDirectory: true)
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        var database: OpaquePointer?
        guard sqlite3_open(root.appendingPathComponent(name).path, &database) == SQLITE_OK,
              let database else {
            throw SQLiteError("Unable to open SQLite database")
        }
        try execute(database, "PRAGMA journal_mode=WAL")
        try execute(database, "PRAGMA synchronous=NORMAL")
        try execute(database, "PRAGMA foreign_keys=ON")
        try execute(database, "PRAGMA busy_timeout=5000")
        try execute(database, "PRAGMA temp_store=MEMORY")
        databases[name] = database
        return database
    }

    private func execute(_ database: OpaquePointer, _ sql: String) throws {
        guard sqlite3_exec(database, sql, nil, nil, nil) == SQLITE_OK else {
            throw SQLiteError(String(cString: sqlite3_errmsg(database)))
        }
    }

    private func prepare(_ database: OpaquePointer, _ sql: String) throws -> OpaquePointer {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(database, sql, -1, &statement, nil) == SQLITE_OK,
              let statement else {
            throw SQLiteError(String(cString: sqlite3_errmsg(database)))
        }
        return statement
    }

    private func decodeArguments(_ json: String) throws -> [Any] {
        let value = try JSONSerialization.jsonObject(with: Data(json.utf8))
        guard let arguments = value as? [Any] else {
            throw SQLiteError("SQLite arguments must be an array")
        }
        return arguments
    }

    private func decodeArgumentSets(_ json: String) throws -> [[Any]] {
        let value = try JSONSerialization.jsonObject(with: Data(json.utf8))
        guard let argumentSets = value as? [[Any]],
              (1...10_000).contains(argumentSets.count) else {
            throw SQLiteError("SQLite executeMany requires between 1 and 10000 argument sets")
        }
        return argumentSets
    }

    private func executeMany(
        _ database: OpaquePointer,
        _ sql: String,
        _ argumentSets: [[Any]]
    ) throws {
        let statement = try prepare(database, sql)
        defer { sqlite3_finalize(statement) }
        try execute(database, "BEGIN IMMEDIATE")
        do {
            for arguments in argumentSets {
                sqlite3_reset(statement)
                sqlite3_clear_bindings(statement)
                try bind(arguments, to: statement)
                guard sqlite3_step(statement) == SQLITE_DONE else {
                    throw SQLiteError(String(cString: sqlite3_errmsg(database)))
                }
            }
            try execute(database, "COMMIT")
        } catch {
            try? execute(database, "ROLLBACK")
            throw error
        }
    }

    private func bind(_ arguments: [Any], to statement: OpaquePointer) throws {
        for (offset, value) in arguments.enumerated() {
            let index = Int32(offset + 1)
            let result: Int32
            switch value {
            case is NSNull:
                result = sqlite3_bind_null(statement, index)
            case let value as Bool:
                result = sqlite3_bind_int64(statement, index, value ? 1 : 0)
            case let value as NSNumber:
                result = sqlite3_bind_double(statement, index, value.doubleValue)
            case let value as String:
                result = sqlite3_bind_text(statement, index, value, -1, SQLITE_TRANSIENT)
            default:
                throw SQLiteError("SQLite arguments must be scalar")
            }
            guard result == SQLITE_OK else { throw SQLiteError("Unable to bind SQLite argument") }
        }
    }

    private func query(
        _ database: OpaquePointer,
        _ sql: String,
        _ arguments: [Any]
    ) throws -> [[String: Any]] {
        let statement = try prepare(database, sql)
        defer { sqlite3_finalize(statement) }
        try bind(arguments, to: statement)
        var rows: [[String: Any]] = []
        while sqlite3_step(statement) == SQLITE_ROW {
            guard rows.count < 1_000 else {
                throw SQLiteError("SQLite query exceeded the 1000-row bridge limit; paginate the query")
            }
            guard sqlite3_column_count(statement) <= 256 else {
                throw SQLiteError("SQLite query exceeded the 256-column bridge limit")
            }
            var row: [String: Any] = [:]
            for index in 0..<sqlite3_column_count(statement) {
                let name = String(cString: sqlite3_column_name(statement, index))
                switch sqlite3_column_type(statement, index) {
                case SQLITE_INTEGER:
                    row[name] = sqlite3_column_int64(statement, index)
                case SQLITE_FLOAT:
                    row[name] = sqlite3_column_double(statement, index)
                case SQLITE_TEXT:
                    row[name] = String(cString: sqlite3_column_text(statement, index))
                case SQLITE_BLOB:
                    let count = Int(sqlite3_column_bytes(statement, index))
                    let bytes = sqlite3_column_blob(statement, index)
                    row[name] = bytes.map { Data(bytes: $0, count: count).base64EncodedString() } ?? ""
                default:
                    row[name] = NSNull()
                }
            }
            rows.append(row)
        }
        return rows
    }

    func close() {
        queue.sync {
            databases.values.forEach { database in
                _ = sqlite3_close(database)
            }
            databases.removeAll()
        }
    }
}

private struct SQLiteError: LocalizedError {
    let message: String
    init(_ message: String) { self.message = message }
    var errorDescription: String? { message }
}

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

File: android/app/src/main/java/dev/pam/nativeapp/modules/SQLiteModule.kt
package dev.pam.nativeapp.modules

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteStatement
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.concurrent.Executors

internal class SQLiteModule(private val context: Context) : NativeModule, AutoCloseable {
    private val executor = Executors.newSingleThreadExecutor()
    private val databases = mutableMapOf<String, SQLiteDatabase>()

    override fun invoke(method: String, payload: ByteArray, completion: ModuleCompletion) {
        executor.execute {
            runCatching {
                val values = WireMap.decode(payload)
                val database = open(values.requiredText("database"))
                val sql = values.requiredText("sql")
                when (method) {
                    "execute" -> {
                        val arguments = decodeArguments(values.requiredText("arguments"))
                        database.execSQL(sql, arguments)
                        completion.complete(ModuleResultStatus.SUCCESS, ByteArray(0))
                    }
                    "query" -> {
                        val arguments = decodeArguments(values.requiredText("arguments"))
                        val rows = database.rawQuery(sql, arguments.map(::stringArgument).toTypedArray())
                            .use(::encodeRows)
                        completion.complete(
                            ModuleResultStatus.SUCCESS,
                            WireMap.encode(mapOf("rows" to WireValue.Text(rows.toString()))),
                        )
                    }
                    "executeMany" -> {
                        executeMany(
                            database,
                            sql,
                            decodeArgumentSets(values.requiredText("arguments")),
                        )
                        completion.complete(ModuleResultStatus.SUCCESS, ByteArray(0))
                    }
                    "transaction" -> {
                        executeTransaction(
                            database,
                            decodeStatements(values.requiredText("arguments")),
                        )
                        completion.complete(ModuleResultStatus.SUCCESS, ByteArray(0))
                    }
                    else -> error("Unknown SQLite method $method")
                }
            }.onFailure { error ->
                completion.complete(
                    ModuleResultStatus.FAILURE,
                    (error.message ?: "SQLite operation failed").toByteArray(),
                )
            }
        }
    }

    private fun open(name: String): SQLiteDatabase =
        databases.getOrPut(name) {
            val root = File(context.filesDir, "pam-databases").apply { mkdirs() }
            SQLiteDatabase.openOrCreateDatabase(File(root, name), null).apply {
                enableWriteAheadLogging()
                setForeignKeyConstraintsEnabled(true)
                pragma(this, "synchronous=NORMAL")
                pragma(this, "busy_timeout=5000")
                pragma(this, "temp_store=MEMORY")
            }
        }

    private fun pragma(database: SQLiteDatabase, expression: String) {
        database.rawQuery("PRAGMA $expression", emptyArray()).use { cursor ->
            cursor.moveToFirst()
        }
    }

    private fun decodeArguments(value: String): Array<Any?> {
        val array = JSONArray(value)
        return Array(array.length()) { index ->
            when (val item = array.opt(index)) {
                JSONObject.NULL -> null
                is Boolean -> if (item) 1L else 0L
                is String, is Number -> item
                else -> error("SQLite arguments must be scalar")
            }
        }
    }

    private fun decodeArgumentSets(value: String): List<Array<Any?>> {
        val batches = JSONArray(value)
        require(batches.length() in 1..MAX_BATCH_ROWS) {
            "SQLite executeMany requires between 1 and 10000 argument sets"
        }
        return List(batches.length()) { index ->
            decodeArguments(batches.getJSONArray(index).toString())
        }
    }

    private fun executeMany(
        database: SQLiteDatabase,
        sql: String,
        argumentSets: List<Array<Any?>>,
    ) {
        val statement = database.compileStatement(sql)
        database.beginTransactionNonExclusive()
        try {
            argumentSets.forEach { arguments ->
                statement.clearBindings()
                bind(arguments, statement)
                statement.execute()
            }
            database.setTransactionSuccessful()
        } finally {
            database.endTransaction()
            statement.close()
        }
    }

    private fun decodeStatements(value: String): List<Pair<String, Array<Any?>>> {
        val statements = JSONArray(value)
        require(statements.length() in 1..MAX_BATCH_ROWS) {
            "SQLite transaction requires between 1 and 10000 statements"
        }
        return List(statements.length()) { index ->
            val statement = statements.getJSONObject(index)
            val sql = statement.getString("sql")
            require(sql.isNotEmpty() && sql.toByteArray().size <= MAX_SQL_BYTES) {
                "Invalid SQLite transaction SQL statement"
            }
            sql to decodeArguments(statement.getJSONArray("arguments").toString())
        }
    }

    private fun executeTransaction(
        database: SQLiteDatabase,
        statements: List<Pair<String, Array<Any?>>>,
    ) {
        database.beginTransactionNonExclusive()
        try {
            statements.forEach { (sql, arguments) ->
                database.compileStatement(sql).use { statement ->
                    bind(arguments, statement)
                    statement.execute()
                }
            }
            database.setTransactionSuccessful()
        } finally {
            database.endTransaction()
        }
    }

    private fun bind(arguments: Array<Any?>, statement: SQLiteStatement) {
        arguments.forEachIndexed { offset, value ->
            val index = offset + 1
            when (value) {
                null -> statement.bindNull(index)
                is Boolean -> statement.bindLong(index, if (value) 1L else 0L)
                is Byte, is Short, is Int, is Long ->
                    statement.bindLong(index, (value as Number).toLong())
                is Float, is Double -> statement.bindDouble(index, (value as Number).toDouble())
                is String -> statement.bindString(index, value)
                else -> error("SQLite arguments must be scalar")
            }
        }
    }

    private fun stringArgument(value: Any?): String =
        when (value) {
            null -> ""
            is Boolean -> if (value) "1" else "0"
            else -> value.toString()
        }

    private fun encodeRows(cursor: Cursor): JSONArray =
        JSONArray().apply {
            while (cursor.moveToNext()) {
                require(length() < MAX_QUERY_ROWS) {
                    "SQLite query exceeded the 1000-row bridge limit; paginate the query"
                }
                require(cursor.columnCount <= MAX_QUERY_COLUMNS) {
                    "SQLite query exceeded the 256-column bridge limit"
                }
                put(JSONObject().apply {
                    repeat(cursor.columnCount) { index ->
                        put(
                            cursor.getColumnName(index),
                            when (cursor.getType(index)) {
                                Cursor.FIELD_TYPE_NULL -> JSONObject.NULL
                                Cursor.FIELD_TYPE_INTEGER -> cursor.getLong(index)
                                Cursor.FIELD_TYPE_FLOAT -> cursor.getDouble(index)
                                Cursor.FIELD_TYPE_BLOB ->
                                    android.util.Base64.encodeToString(cursor.getBlob(index), android.util.Base64.NO_WRAP)
                                else -> cursor.getString(index)
                            },
                        )
                    }
                })
            }
        }

    override fun close() {
        executor.shutdown()
        databases.values.forEach(SQLiteDatabase::close)
        databases.clear()
    }

    private fun Map<String, WireValue>.requiredText(key: String): String =
        (this[key] as? WireValue.Text)?.value ?: error("Missing text field $key")

    private companion object {
        const val MAX_QUERY_ROWS = 1_000
        const val MAX_QUERY_COLUMNS = 256
        const val MAX_BATCH_ROWS = 10_000
        const val MAX_SQL_BYTES = 1_048_576
    }
}
